providers/default.rb

```diff
@@ -0,0 +1,157 @@
+#
+# Cookbook Name:: python-webapp
+# Provider:: default
+#
+# Copyright 2015, Oregon State University
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+action :install do
+  # Use the git recipe to install git
+  run_context.include_recipe 'git'
+  run_context.include_recipe 'python::pip'
+  run_context.include_recipe 'python::virtualenv'
+
+  # required to work around https://github.com/poise/python/issues/100
+  python_pip 'setuptools' do
+    action :install
+    options '--upgrade'
+  end
+
+  if new_resource.create_user
+    group new_resource.group do
+      action :create
+    end
+    user new_resource.owner do
+      action :create
+      gid new_resource.group
+    end
+  end
+
+  # If the new_resource.path is nil set the install path to
+  # `/opt/name_attribute`
+  path = new_resource.path || "/opt/#{ new_resource.name }"
+
+  if new_resource.virtualenv_path.nil?
+    virtualenv_path = "/opt/venv_#{ new_resource.name }"
+  else
+    virtualenv_path = new_resource.virtualenv_path
+  end
+
+  # If the config_destination is nil, as by default, default to
+  # /opt/<path>/settings.py
+  if new_resource.config_destination.nil?
+    config_destination = "#{ path }/settings.py"
+  else
+    config_destination = new_resource.config_destination
+  end
+
+  directory path do
+    action :create
+    recursive true
+    owner new_resource.owner
+    group new_resource.group
+  end
+
+  # Create virtual environment
+  python_virtualenv virtualenv_path do
+    interpreter new_resource.interpreter
+    owner new_resource.owner
+    group new_resource.group
+    action :create
+  end
+
+  # Update the code.
+  git path do
+    action :sync
+    repository new_resource.repository
+    checkout_branch new_resource.revision
+    destination path
+    user new_resource.owner
+    group new_resource.group
+  end
+
+  # If a config file template has been specified, create it.
+  template new_resource.config_template do
+    only_if { !new_resource.config_template.nil? }
+    action :create
+    source new_resource.config_template
+    path config_destination
```

Comment:
Actually, it appears doing this would be non-trivial (as you'd have to pass in the directory and filenames separately, or separate them with regexfoo). We should leave this up to the user.